src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs

```diff
@@ -719,21 +710,30 @@ private static void EscapeAsciiChar(char ch, char[] to, ref int pos)
         /// </summary>
         /// <param name="path"></param>
         /// <returns></returns>
-        private static string EscapeSpecialCharacters(string path)
+        private static unsafe string EscapeSpecialCharacters(string path)
         {
-            string characterString;
+            var builder = new StringBuilder(path.Length);
+            const int charsLength = 3;
+            char* chars = stackalloc char[charsLength];
 
             // Escaping for the following - '%'; '@'; ',' and '?'
             // !!Important!! - The order is important - The '%' sign should be escaped first.
             // This is currently enforced by the order of characters in the _specialCharacters array
-            foreach (char c in s_specialCharacters)
+            foreach (char c in path)
             {
-                characterString = c.ToString();
-                if (path.Contains(characterString))
-                    path = path.Replace(characterString, HexEscape(c));
+                if (Array.IndexOf(s_specialCharacters, c) < 0)
+                {
+                    builder.Append(c);
+                }
+                else
+                {
+                    int pos = 0;
+                    EscapeAsciiChar(c, chars, ref pos);
+                    builder.Append(chars, charsLength);
+                }
```

Comment:
This transformation is not adhering to the comment on the loop:
```C#
// !!Important!! - The order is important - The '%' sign should be escaped first.
```
While it could be made more efficient than this, as a compromise between efficiency and simplicity, I'd suggest something like this:
```C#
if (!path.IndexOfAnyChar(s_specialCharacters))
{
    return path;
}

var builder = new StringBuilder(path);
foreach (char c in s_specialCharacters)
{
    if (path.Contains(c))
    {
        builder.Replace(c.ToString(), HexEscape(c));
    }
}
return builder.ToString();
```
This way, we at least avoid any allocations if there's nothing to replace, we avoid allocating an entirely new path string each time something is replaced, etc.  We do still allocate strings for the char c and the escaped char c, but that could be easily avoided if desired by having those strings precalculated, e.g.
```C#
private readonly static char[] s_specialCharacters = ...;
private readonly static string[] s_specialCharactersAsStrings = ...;
private readonly static string[] s_specialCharactersHexEncoded = ...;
```
and then using the appropriate cached strings from here rather than allocating them each time.